### PR TITLE
Feature/add pypy support

### PR DESCRIPTION
**Pull request by snudler6**

### CODE REVIEW

File: src/time_travel/patchers/datetime_patcher.py [hunk 1]
```diff
@@ -1,30 +1,246 @@
...
+        
+        for module in modules:
+            for attr in dir(module):
+                try:
+                    attribute_value = getattr(module, attr)
+                except (ValueError, AttributeError, ImportError):
+                    # For some libraries, this happen.
+                    continue
+                
+                if attr in local_names or\
```
Comment (FarmerSez) [inline src/time_travel/patchers/datetime_patcher.py]: Space between `or` and `\`.

File: src/time_travel/patchers/datetime_patcher.py [hunk 2]
```diff
@@ -1,30 +1,246 @@
...
     
     patching:
         - datetime.today
+        - datetime.now
+        - datetime.utcnow
+        - date.today
     """
     
-    def __init__(self, *args, **kwargs):
+    def __init__(self, patched_modules=None, **kwargs):
```
Comment (FarmerSez) [inline src/time_travel/patchers/datetime_patcher.py]: I really don't like this interface. The patcher is dynamically loaded (using entry points), right? So what tells the user that he has to pass this argument via `kwargs` for the patcher to work other than the docs? It seems like a really weak spot for the package.

File: src/time_travel/patchers/datetime_patcher.py [hunk 3]
```diff
@@ -1,30 +1,246 @@
...
+def pickle_fake_date(datetime_):
+    """Pickle function for FakeDate."""
+    return FakeDate, (
+        datetime_.year,
+        datetime_.month,
+        datetime_.day,
+    )
+
+
+@staticmethod
```
Comment (FarmerSez) [inline src/time_travel/patchers/datetime_patcher.py]: Is this decorator needed? Does it even make any sense?

File: src/time_travel/patchers/datetime_patcher.py [hunk 4]
```diff
@@ -1,30 +1,246 @@
...
+        to_patch = (
+            # (local_name, orig_local_class, fake_name, fake_class)
+            ('_real_date', _real_date, 'FakeDate', FakeDate),
+            ('_real_datetime', _real_datetime, 'FakeDatetime', FakeDatetime),
+        )
+        
+        local_names = tuple(real_name for real_name, _, _, _ in to_patch)
+        real_id_to_fake = {id(real): fake for _, real, _, fake in to_patch}
+
+        modules = [sys.modules[name] for name in self.patched_modules] if \
```
Comment (FarmerSez) [inline src/time_travel/patchers/datetime_patcher.py]: This is completely unreadable.

File: src/time_travel/patchers/datetime_patcher.py [hunk 5]
```diff
@@ -1,30 +1,246 @@
...
+                
+    def stop(self):
+        """Stop the patching of datetime module."""
+        datetime.date = _real_date
+        datetime.datetime = _real_datetime
+        
+        for module, attribute, original_value in self._undo_set:
+            setattr(module, attribute, original_value)
+            
+        self._undo_set = set()
```
Comment (FarmerSez) [inline src/time_travel/patchers/datetime_patcher.py]: `self._undo_set.clear()`

File: src/time_travel/patchers/datetime_patcher.py [hunk 6]
```diff
@@ -1,30 +1,246 @@
...
+            self.patched_modules = patched_modules
+        else:
+            self.patched_modules = [patched_modules]
+        
+        self._undo_set = set()
         
-        self.datetime = mock.Mock(wraps=datetime_lib.datetime)
-        self.datetime.mock_add_spec(
-            ["today", "now", "utcnow", "fromtimestamp"])
+    def start(self):
```
Comment (FarmerSez) [inline src/time_travel/patchers/datetime_patcher.py]: This method needs **a lot** more documentation. It's impossible to understand the thought process as it is now.

File: src/time_travel/patchers/datetime_patcher.py (hunk 2 above)
Comment (FarmerSez) [reply src/time_travel/patchers/datetime_patcher.py]: Also, what happens if it's None? Nothing is patched?

Comment (snudler6) [reply src/time_travel/patchers/datetime_patcher.py]: If It's `None`, The code will pass on all modules, and try to find any module using datetime and replace it.
If it's not None, The code will search datetime only in the  given modules

Comment (snudler6) [reply src/time_travel/patchers/datetime_patcher.py]: Moved it to the basic patcher and opened an issue (#24) to change all patchers to use the patched_modules arg.

File: src/time_travel/patchers/datetime_patcher.py [hunk 7]
```diff
@@ -202,28 +195,43 @@ def start(self):
...
+            # Patch on all loaded modules
+            modules = [
                 module for mod_name, module in sys.modules.items() if
-                mod_name is not None and module is not None and
+                mod_name is not None and
+                module is not None and
                 hasattr(module, '__name__') and
+                # Don't patch inside this module,
+                # or inside the original module.
                 module.__name__ not in ('datetime', 'datetime_patcher')
```
Comment (FarmerSez) [inline src/time_travel/patchers/datetime_patcher.py]: Is it possible to do `module.__name__ not in ('datetime', __name__)`?

File: README.md [hunk 8]
```diff
@@ -8,7 +8,9 @@ python time libraries mocking
```
Comment (FarmerSez) [inline README.md]: Line 7 (can't comment on it): CAN HAS POTATO.

File: README.md [hunk 9]
```diff
@@ -8,7 +8,9 @@ python time libraries mocking
 
 ## Usage
 
-`TimeTravel` is context manager patching all time modules currently patched in a single line.
+`TimeTravel` is context manager patching all* time and I\O related modules in a single line.
+
+* All modules currently patched :) .
```
Comment (FarmerSez) [inline README.md]: This will look like a bullet on the generated page.

- All modules ...

File: src/time_travel/patchers/basic_patcher.py [hunk 10]
```diff
@@ -4,11 +4,22 @@
 class BasicPatcher(object):
     """Base class for patching time modules."""
     
-    def __init__(self, clock, events_pool=None, **kwargs):
+    def __init__(self,
+                 clock,
+                 events_pool=None,
+                 patched_modules=None,
```
Comment (FarmerSez) [inline src/time_travel/patchers/basic_patcher.py]: I'd call that `modules_to_patch`.

File: README.md (hunk 8 above)
Comment (snudler6) [reply README.md]: image

File: src/time_travel/patchers/datetime_patcher.py (hunk 7 above)
Comment (snudler6) [reply src/time_travel/patchers/datetime_patcher.py]: cools.

File: README.md (hunk 9 above)
Comment (snudler6) [reply README.md]: Oops.

File: src/time_travel/patchers/basic_patcher.py (hunk 10 above)
Comment (snudler6) [reply src/time_travel/patchers/basic_patcher.py]: Agreed.

File: README.md [hunk 11]
```diff
@@ -4,13 +4,13 @@
 python time libraries mocking
 
 ### time-travel is the fun and easy way to unit-test time sensetive modules.
-### Can write determenistic test for long and even infinite time scenarios.
+### time-travel let you write determenistic test for long and even infinite time scenarios.
```
Comment (FarmerSez) [inline README.md]: lets 🙄

Comment (snudler6) [reply README.md]: Boo

**Pull request merged**